modules/core/04-channel/types/upgrade.go

```diff
@@ -77,12 +71,37 @@ func (ut Timeout) IsValid() bool {
 	return !ut.Height.IsZero() || ut.Timestamp != 0
 }
 
-// NewErrorReceipt returns an error receipt with the code from the provided error type stripped
-// out to ensure changes of the error message don't cause state machine breaking changes.
-func NewErrorReceipt(upgradeSequence uint64, err error) ErrorReceipt {
-	_, code, _ := errorsmod.ABCIInfo(err, false) // discard non-determinstic codespace and log values
+// UpgradeError defines an error that occurs during an upgrade.
+type UpgradeError struct {
+	// err is the underlying error that caused the upgrade to fail.
+	// this error should not be written to state.
+	err error
+	// sequence is the upgrade sequence number of the upgrade that failed.
+	sequence uint64
+}
+
+// NewUpgradeError returns a new UpgradeError instance.
+func NewUpgradeError(upgradeSequence uint64, err error) UpgradeError {
+	return UpgradeError{
+		err:      err,
+		sequence: upgradeSequence,
+	}
+}
+
+// Error implements the error interface, returning the underlying error which caused the upgrade to fail.
+func (u UpgradeError) Error() string {
+	return u.err.Error()
+}
+
+// GetErrorReceipt returns an error receipt with the code from the underlying error type stripped.
+func (u UpgradeError) GetErrorReceipt() ErrorReceipt {
```

Comment:
we could add a small unit test for this I guess

Reply:
maybe assert that 2 of the same base errors wrapped differently still return the same error receipt?

Reply:
hmm, interesting. If the same base error is wrapped differently will we get the code from the base error or the top level

Reply:
A wrapped error should only have 1 code (the base error), wrapping just adds a string to the `wrappedError.description`

Reply:
Yep, we had to modify unwrapping such that the base error code is always returned. When we started writing a test we could see a different code returned in `GetErrorReceipt` as it was taking the top level code as opposed to the base error code

Reply:
Yeah, the issue was when wrapping the `UpgradeError` itself, as `Unwrap` wasn't implemented like it is for errors registered with errorsmod.

Reply:
For example this is fine:

```go
upgradeErr := channeltypes.NewUpgradeError(1, channeltypes.ErrInvalidChannel)
wrappedUpgradeErr := channeltypes.NewUpgradeError(1, errorsmod.Wrap(channeltypes.ErrInvalidChannel, "testing"))

suite.Require().Equal(upgradeErr.GetErrorReceipt().Message, wrappedUpgradeErr.GetErrorReceipt().Message)
```

But this isn't (without `Unwrap` implemented on `UpgradeError`):

```go
ue := channeltypes.NewUpgradeError(1, channeltypes.ErrInvalidChannel)
wrappedErr := errorsmod.Wrap(ue, "wrapped upgrade error")
suite.Require().True(errorsmod.IsOf(wrappedErr, channeltypes.ErrInvalidChannel))

ue2 := channeltypes.NewUpgradeError(1, wrappedErr)
suite.Require().Equal(ue2.GetErrorReceipt().Message, ue.GetErrorReceipt().Message)
```

Reply:
lol, see https://github.com/cosmos/ibc-go/pull/3700#discussion_r1213197560, we can close this thread